File: fastapi_migration/backend/app/schemas/dna.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, List, Any
from datetime import datetime

# DNA file schemas
class DNAFileBase(BaseModel):
    filename: str
    file_hash: str

class DNAFileCreate(DNAFileBase):
    file_path: Optional[str] = None
    user_id: Optional[int] = None
    file_format: Optional[str] = None
    snp_count: Optional[int] = None

class DNAFileUpdate(BaseModel):
    file_path: Optional[str] = None
    file_format: Optional[str] = None
    snp_count: Optional[int] = None
    status: Optional[str] = None
    error_message: Optional[str] = None

class DNAFileInDBBase(DNAFileBase):
    id: int
    status: str
    created_at: datetime
    file_format: Optional[str] = None
    snp_count: Optional[int] = None
    error_message: Optional[str] = None
    
    class Config:
        from_attributes = True

class DNAFile(DNAFileInDBBase):
    """DNA file model returned to client"""
    pass

# API response schemas
class DNAFileResponse(BaseModel):
    filename: str
    file_hash: str
    status: str
    snp_count: Optional[int] = None
    processing_time: Optional[float] = None
    message: Optional[str] = None
    cached: bool = False

class SNPDataResponse(BaseModel):
    filename: str
    valid: bool
    format: Optional[str] = None
    snp_count: int = 0
    chromosomes: Dict[str, int] = {}
    statistics: Dict[str, Any] = {}
    errors: List[str] = []

# SNP-related schemas
class SNP(BaseModel):
    rsid: str
    chromosome: str
    position: str
    allele1: str
    allele2: str

class SNPBatch(BaseModel):
    snps: List[SNP] = []

File: fastapi_migration/frontend/src/types/dna.ts
/**
 * Types related to DNA data and processing
 */

export interface SNPID {
  rsid: string;
}

export interface SNP extends SNPID {
  chromosome: string;
  position: string;
  allele1: string;
  allele2: string;
}

export interface DNAFileUploadResponse {
  file_name: string;
  file_hash: string;
  file_size: number;
  upload_date: string;
  status: string;
  snp_count?: number;
  processing_time?: number;
  message?: string;
  cached: boolean;
  format?: string;
}

export interface ValidationStatistics {
  alleles?: Record<string, number>;
  rsid_patterns?: Record<string, number>;
}

export interface ValidationResult {
  valid: boolean;
  format?: string;
  snp_count: number;
  chromosomes: Record<string, number>;
  statistics: ValidationStatistics;
  errors: string[];
}

export interface DNAFile {
  id?: number;
  file_hash: string;
  filename: string;
  file_format?: string;
  snp_count?: number;
  status: string;
  created_at?: string;
}

export interface ValidatedFile extends DNAFile {
  validation: ValidationResult;
}

export interface DNAUploadProgress {
  uploaded: boolean;
  progress: number;
  validated: boolean;
  processed: boolean;
  error?: string;
}

export interface FileFormat {
  name: string;
  description: string;
  columns?: string[];
  example?: string;
  example_line?: string;
  file_pattern?: string;
  support_level: 'full' | 'partial' | 'experimental';
}